### ISSUE TITLE
error: Failed to translate a method call.: InvalidArg(..) [Microsoft.FSharp.Core.Operators]

### ISSUE THREAD

**Issue by Martin521** [intellifactory/websharper#721]
It seems `invalidArg` is not supported.
I can work around it, but would be nice to have this basic F# feature supported. Or is there a reason not to?
(same for invalidOp)

**Comment by Martin521** [comment]
I am coming across more method calls that WebSharper fails to translate. Is there any documentation that tells me what is allowed and what not?

**Comment by granicz** [comment]
No documentation as of yet, but we are planning to add an "API lookup" facility on the [WebSharper developer site](http://developers.websharper.com) where you type a FQN to see if it can be used in client-side code. Would this be useful in your case? We might even make this a CLI tool as well for offline use.

As for those failing methods, please add them on this ticket and we will include them in the next release.

**Comment by cgravill** [comment]
Having a tool we could point at a given .net project and getting a list of what's supported or not would be really helpful. We have some more potential tools that we might want to use with WebSharper with but we don't have an estimate of how possible it is. If it's just a few methods we could potentially add them ourselves and submit a pull request but we'd rather not get halfway and find out there's lots missing!

**Comment by granicz** [comment]
You can always try a simple trick and add `[<JavaScript>]` to all of your modules, and compile the code inside a WebSharper project.  This will yield warnings/errors for any method that is not yet supported for client-side use.

**Comment by cgravill** [comment]
Yep, I've thought about that. However, with multiple projects and tens of files in each it's a fair amount of mechanical effort. There's also the issue of duplicate errors. I'm guessing going between projects wouldn't work - if project A fails then project B depending on A will have missing methods? All this could be post-processed out but a tool to assess this would be great.

**Comment by Martin521** [comment]
@granicz : That's exactly what I am doing now.
But coming back to my original question: Are there any plans to add the missing methods?
I can easily work around `invalidArg` (as `failwith` is supported). But others (such as `GetLength()` of multi-dimensional arrays; or `GetEnumerator()` of `IEnumerable`) are difficult to replace.

**Comment by Martin521** [comment]
Here are some methods that I can work around, but would be nice to have:
`Operators.invalidArg`
`Operators.invalidOp`
`Array.GetLength`
`File.ReadAllLines`
`File.WriteAllLines`

And here are two remaining issues that would need more re-writing in my code. But perhaps are also difficult to translate.
Extending a type to `IEnumerable` by supplying `IEnumerable.GetEnumerator` (=> Error "failed to translate a method call")
Instantiating `TimeSpan` (=> Websharper complains about static field `Zero`, which I don't use)

**Comment by semuserable** [comment]
That's exactly what I ran into myself, some methods just don't work on client side.
The one I'm looking into right now is `DateTime.ParseExact`.

**Comment by Martin521** [comment]
@Jand42 Thanks for implementing this so promptly!

**Comment by Jand42** [comment]
@Martin521 sorry, these are still not back-ported to WS3, I added a label to indicate ticket does not refer to current master which is WS4 beta.

**Comment by Martin521** [comment]
I am using now WS4 beta, so no problem for me :-)

**Comment by granicz** [comment]
@Martin521 @cgravill @semuserable Re: the API browser/lookup facility I mentioned, here is the current state we have under development:

![screenshot 2017-08-11 03 55 15](https://user-images.githubusercontent.com/6954755/29199199-09095f22-7e4b-11e7-8a2c-ae5df8845d52.png)

It contains a searchable tree of types defined in core .NET and F# libraries, and indicates which types contain proxies. The documentation page for any given type then lists all constructors, members, and properties with an indication whether they can be used in client-side code.

Can you think of ways to extend this functionality?  We already found it suitable to locate gaps in proxy coverage.

**Comment by Martin521** [comment]
Yes, this will be a valuable (today really missing) part of the documentation.
I have been able to work around the few missing functions that I encountered, but in the beginning I was completely in the dark if I had make a mistake or if the function was missing.
I am looking forward to having this on websharper.com.
No immediate wishes for extended functionality.
BTW after working with Websharper now a little (and really enjoying it) I am less bothered by _missing_ functionality than by _incompatible_ functionality (see my other recent posts).
